nameserver/dns.go

```diff
@@ -96,6 +99,40 @@ func (d *DNSServer) listen(address string) error {
 	return nil
 }
 
+func (d *DNSServer) readUpstream() *dns.ClientConfig {
+	d.upstreamMutex.Lock()
+	defer d.upstreamMutex.Unlock()
+
+	// If we checked less than five seconds ago return what we already have
+	now := time.Now()
+	d.ns.debugf("Reading upstream info (now = %v, last = %v)", now, d.upstreamChecked)
+	if now.Sub(d.upstreamChecked) < 5*time.Second {
+		d.ns.debugf("Returning cached value %v", d.upstream)
+		return d.upstream
+	}
+	d.upstreamChecked = now
+
+	// Reread file if it has changed
+	if fi, err := os.Stat(etcResolvConf); err == nil {
+		d.ns.debugf("Stat %v (was %v, now %v)", etcResolvConf, d.upstreamModified, fi.ModTime())
+		if fi.ModTime() != d.upstreamModified {
+			if upstream, err := dns.ClientConfigFromFile(etcResolvConf); err == nil {
+				upstream.Servers = filter(upstream.Servers, d.effectiveAddress)
+				d.upstream = upstream
+				d.upstreamModified = fi.ModTime()
+				d.ns.debugf("Read new %v: %v", etcResolvConf, d.upstream)
+			} else {
+				d.ns.debugf("Read error: %v; returning existing cached value %v", err, d.upstream)
+			}
+		} else {
+			d.ns.debugf("Unmodified; returning existing cached value %v", d.upstream)
+		}
+	} else {
+		d.ns.debugf("Stat error: %v; returning existing cached value %v", err, d.upstream)
+	}
+	return d.upstream
+}
```

Comment:
Indeed - I wrote it that way initially, and then I had to figure out why it wasn't working :smile: Do you really want them removed? They'll be helpful if this breaks again in future...